app/controllers/api_controller/virtual_templates.rb

```diff
@@ -0,0 +1,26 @@
+class ApiController
+  module VirtualTemplates
+    def create_resource_virtual_templates(type, _id, data)
+      validate_vendor_present(data)
+      klass = collection_class(type)
+      vt_type = type_from_vendor(klass, data['vendor'])
+      template = klass.create(data.merge(:type => vt_type))
+      if template.invalid?
+        raise BadRequestError, "Failed to create a new virtual template - #{template.errors.full_messages.join(', ')}"
+      end
+      template
+    end
+
+    private
+
+    def validate_vendor_present(data)
+      raise BadRequestError, 'Must specify a vendor for creating a Virtual Template' unless data['vendor']
+    end
+
+    def type_from_vendor(klass, vendor)
+      type = klass::TYPES[vendor.to_sym]
```

Comment:
WDYT about doing this in a single line:

```ruby
klass::TYPES.fetch(vendor.to_sym) { raise "etc..." }
```

Yours seems clearer though, so not necessary